src/common/src/array/column.rs

```diff
@@ -53,6 +54,39 @@ impl Column {
     pub fn into_inner(self) -> ArrayRef {
         self.array
     }
+
+    pub fn expand_columns(
+        cardinality: usize,
+        columns: Vec<Column>,
+        column_subsets: Vec<Vec<usize>>,
+    ) -> ArrayResult<Vec<Vec<Column>>> {
+        let null_columns: Vec<Column> = columns
+            .iter()
+            .map(|column| {
+                let array = column.array_ref();
+                let mut builder = array.create_builder(cardinality)?;
+                // TODO: use a more efficient way to generate `null_column`.
+                (0..cardinality).try_for_each(|_i| builder.append_null())?;
+                Ok::<Column, ArrayError>(Column::new(Arc::new(builder.finish()?)))
+            })
+            .try_collect()?;
+        column_subsets
+            .into_iter()
+            .enumerate()
+            .map(|(i, subset)| {
+                let mut new_columns = null_columns.clone();
+                for key in subset {
+                    new_columns[key] = columns[key].clone();
+                }
+                let flags = Column::from(PrimitiveArray::<i64>::from_slice(&vec![
+                    Some(i as i64);
+                    cardinality
+                ])?);
+                new_columns.push(flags);
+                Ok::<Vec<Column>, ArrayError>(new_columns)
+            })
+            .try_collect()
```

Comment:
Similarly, it is better to yield the output of each subset rather than processing all subsets at once.

However, it would be unnecessarily complex to yield inside this function. We can keep this as a normal function that only input/output for one subset. And the shared null_column also built once and passed into each call.